lib/rubocop/cop/root_cops/factory_file_name.rb

```diff
@@ -1,25 +1,67 @@
-require "pry"
+require "active_support/inflector"
 
 module RuboCop
   module Cop
     module RootCops
       class FactoryFileName < Cop
-        FILE_NAME_ERROR = "Factory name should match file name".freeze
+        # Whitelisted words feel like invariant nouns but they are really not
+        # Ex: the singular of data is datum, but we should be able to name a factory '*_data' without error
+        WHITELIST = ["data".freeze, "metadata".freeze].freeze
 
         def investigate(processed_source)
           file_path = processed_source.buffer.name
           @base_file_name = File.basename(file_path, ".rb")
+          @base_file_name_body = _generate_body(@base_file_name)
+          @base_file_name_last_word = @base_file_name.split("_").last
         end
 
         def on_send(node)
-          method_name = node.method_name
+          return if _file_name_has_error?
+
+          receiver, method_name = *node
+          _, receiver_name = *receiver
+
+          if receiver_name == :FactoryBot && method_name == :define
+            unless _file_name_pluralized?
+              add_offense(
+                node,
+                :location => :expression,
+                :severity => :fatal,
+                :message => "Factory file name should be plural (#{@base_file_name.pluralize})."
+              )
+            end
+          end
+
           return unless _factory_method?(method_name)
 
           factory_symbol_name = *node.arguments[0]
-          factory_name = factory_symbol_name[0].id2name.split("__").first
+          factory_name_array = factory_symbol_name[0].id2name.split("__")
+          factory_name = factory_name_array.first
+          factory_name_body = _generate_body(factory_name)
+          factory_name_last_word = factory_name.split("_").last
 
-          if factory_name != @base_file_name
-            add_offense(node, :location => :expression, :message => FILE_NAME_ERROR)
+          if factory_name_array.size > 1
+            if factory_name_array[0] != @base_file_name
+              add_offense(
+                node,
+                :location => :expression,
+                :message => "Factory name uses incorrect prefix, should be '#{@base_file_name}__#{factory_name_array[1]}'."
+              )
+            end
+
+          elsif factory_name == @base_file_name && !_whitelisted?(@base_file_name_last_word) && !_invariant_noun?(@base_file_name_last_word)
```

Comment:
I don't like the `!_whitelisted?` and `!_invariant_noun?` here. Maybe there should be `_not_whitelisted` and `_not_invariant_noun` methods instead?

Reply:
I think that's a good idea. Typically I try to keep methods positive (as you currently have) so that they are more reusable and can be negated easily. However, since we will never be reusing these outside this cop, I think the readability of `... && _not_whitelisted?(@base_file_name_last_word) && ...` is more clear.